### PR TITLE
Close the API

### PR DESCRIPTION
**Pull request by greg0ire**
This allows big signature changes to take place.

I am targetting this branch, because this is major.

## Changelog

```markdown
### Removed
- extending classes from the package is no longer possible
```

### CODE REVIEW

File: src/FixedBitNotation.php [hunk 1]
```diff
@@ -37,18 +37,8 @@ class FixedBitNotation
      * @param bool   $padFinalGroup     Add padding to end of encoded output
      * @param string $padCharacter      Character to use for padding
      */
-    public function __construct(int $bitsPerCharacter, string $chars = null, bool $rightPadFinalBits = false, bool $padFinalGroup = false, string $padCharacter = '=')
+    final public function __construct(int $bitsPerCharacter, string $chars = null, bool $rightPadFinalBits = false, bool $padFinalGroup = false, string $padCharacter = '=')
```
Comment (core23) [inline src/FixedBitNotation.php]: Do we need this on a final class?

Comment (greg0ire) [inline src/FixedBitNotation.php]: Woops. We sure don't I made a mistake.

Comment (jordisala1991) [inline src/FixedBitNotation.php]: There is no need to define it final if the class is already final: http://php.net/manual/en/language.oop5.final.php

Comment (greg0ire) [inline src/FixedBitNotation.php]: I know lol,  it was just a brain fart

**Comment by greg0ire** [comment]
I think after that, we can release 2.0. Can you think of anything left to do before that?

File: src/GoogleAuthenticator.php [hunk 2]
```diff
@@ -136,7 +126,7 @@ public function generateSecret()
      *
      * @return int
      */
-    protected static function hashToInt($bytes, $start)
+    private static function hashToInt($bytes, $start)
```
Comment (jordisala1991) [inline src/GoogleAuthenticator.php]: I think this method should not be static.

**Comment by jordisala1991** [comment]
Maybe typehint this method: `private static function hashToInt($bytes, $start)` but not needed in order to do a release.

File: src/GoogleAuthenticator.php [hunk 3]
```diff
@@ -136,7 +126,7 @@ public function generateSecret()
      *
      * @return int
      */
-    protected static function hashToInt($bytes, $start)
+    private function hashToInt(string $bytes, int $start): int
```
Comment (jordisala1991) [inline src/GoogleAuthenticator.php]: I think we should also change docblock `@param` section also if we doing this typehint change

File: src/FixedBitNotation.php [hunk 4]
```diff
@@ -20,15 +20,15 @@
...
+final class FixedBitNotation
 {
-    protected $_chars;
-    protected $_bitsPerCharacter;
-    protected $_radix;
-    protected $_rightPadFinalBits;
-    protected $_padFinalGroup;
-    protected $_padCharacter;
-    protected $_charmap;
+    private $_chars;
```
Comment (core23) [inline src/FixedBitNotation.php]: We should add PHPdoc here.

**Comment by jordisala1991** [comment]
> I think after that, we can release 2.0. Can you think of anything left to do before that?

We should remove `paragonie/random_compat` from require dependencies. It's only for compatibility with PHP < 7.

Also we should update sample code, its still pointing to old `lib`dir.

File: src/GoogleAuthenticator.php (hunk 3 above)
Comment (greg0ire) [inline src/GoogleAuthenticator.php]: Very true

**Comment by greg0ire** [comment]
Please review again @core23

File: src/FixedBitNotation.php [hunk 5]
```diff
@@ -185,24 +202,24 @@ public function decode($encodedString, $caseSensitive = true, $strict = false)
...
-        $radix = $this->_radix;
-        $rightPadFinalBits = $this->_rightPadFinalBits;
-        $padFinalGroup = $this->_padFinalGroup;
-        $padCharacter = $this->_padCharacter;
+        $chars = $this->chars;
+        $bitsPerCharacter = $this->bitsPerCharacter;
+        $radix = $this->radix;
+        $rightPadFinalBits = $this->rightPadFinalBits;
+        $padFinalGroup = $this->padFinalGroup;
+        $padCharacter = $this->padCharacter;
```
Comment (jordisala1991) [inline src/FixedBitNotation.php]: Is all this needed? I mean, can't we just use $this-> variables?

**Comment by greg0ire** [comment]
@core23 see https://github.com/sonata-project/GoogleAuthenticator/pull/57

**Comment by greg0ire** [comment]
@sonata-project/contributors please review again

**Comment by greg0ire** [comment]
@sonata-project/contributors please merge

**Comment by jordisala1991** [comment]
Squash?

**Comment by greg0ire** [comment]
No, regular merge.

**Comment by jordisala1991** [comment]
Thanks @greg0ire

**Pull request merged**